### PR TITLE
[SPARK-43646][CONNECT][TESTS] Make both SBT and Maven use `spark-proto` uber jar to test the `connect` module

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
Before this pr, when we tested the `connect` module, Maven used the shaded `spark-protobuf` jar for testing, while SBT used the original jar for testing, which also led to inconsistent testing behavior. So some tests passed when using SBT, but failed when using Maven:

run

```
build/mvn clean install -DskipTests
build/mvn test -pl connector/connect/server
```

there will be two test failed as follows:

```
- from_protobuf_messageClassName *** FAILED ***
  org.apache.spark.sql.AnalysisException: [CANNOT_LOAD_PROTOBUF_CLASS] Could not load Protobuf class with name org.apache.spark.connect.proto.StorageLevel. org.apache.spark.connect.proto.StorageLevel does not extend shaded Protobuf Message class org.sparkproject.spark_protobuf.protobuf.Message. The jar with Protobuf classes needs to be shaded (com.google.protobuf.* --> org.sparkproject.spark_protobuf.protobuf.*).
  at org.apache.spark.sql.errors.QueryCompilationErrors$.protobufClassLoadError(QueryCompilationErrors.scala:3417)
  at org.apache.spark.sql.protobuf.utils.ProtobufUtils$.buildDescriptorFromJavaClass(ProtobufUtils.scala:193)
  at org.apache.spark.sql.protobuf.utils.ProtobufUtils$.buildDescriptor(ProtobufUtils.scala:151)
  at org.apache.spark.sql.protobuf.ProtobufDataToCatalyst.messageDescriptor$lzycompute(ProtobufDataToCatalyst.scala:58)
  at org.apache.spark.sql.protobuf.ProtobufDataToCatalyst.messageDescriptor(ProtobufDataToCatalyst.scala:57)
  at org.apache.spark.sql.protobuf.ProtobufDataToCatalyst.dataType$lzycompute(ProtobufDataToCatalyst.scala:43)
  at org.apache.spark.sql.protobuf.ProtobufDataToCatalyst.dataType(ProtobufDataToCatalyst.scala:42)
  at org.apache.spark.sql.catalyst.expressions.Alias.toAttribute(namedExpressions.scala:194)
  at org.apache.spark.sql.catalyst.plans.logical.Project.$anonfun$output$1(basicLogicalOperators.scala:72)
  at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286)


- from_protobuf_messageClassName_options *** FAILED ***
  org.apache.spark.sql.AnalysisException: [CANNOT_LOAD_PROTOBUF_CLASS] Could not load Protobuf class with name org.apache.spark.connect.proto.StorageLevel. org.apache.spark.connect.proto.StorageLevel does not extend shaded Protobuf Message class org.sparkproject.spark_protobuf.protobuf.Message. The jar with Protobuf classes needs to be shaded (com.google.protobuf.* --> org.sparkproject.spark_protobuf.protobuf.*).
  at org.apache.spark.sql.errors.QueryCompilationErrors$.protobufClassLoadError(QueryCompilationErrors.scala:3417)
  at org.apache.spark.sql.protobuf.utils.ProtobufUtils$.buildDescriptorFromJavaClass(ProtobufUtils.scala:193)
  at org.apache.spark.sql.protobuf.utils.ProtobufUtils$.buildDescriptor(ProtobufUtils.scala:151)
  at org.apache.spark.sql.protobuf.ProtobufDataToCatalyst.messageDescriptor$lzycompute(ProtobufDataToCatalyst.scala:58)
  at org.apache.spark.sql.protobuf.ProtobufDataToCatalyst.messageDescriptor(ProtobufDataToCatalyst.scala:57)
  at org.apache.spark.sql.protobuf.ProtobufDataToCatalyst.dataType$lzycompute(ProtobufDataToCatalyst.scala:43)
  at org.apache.spark.sql.protobuf.ProtobufDataToCatalyst.dataType(ProtobufDataToCatalyst.scala:42)
  at org.apache.spark.sql.catalyst.expressions.Alias.toAttribute(namedExpressions.scala:194)
  at org.apache.spark.sql.catalyst.plans.logical.Project.$anonfun$output$1(basicLogicalOperators.scala:72)
  at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286) 
```

So this pr make SBT also use `spark-proto` uber jar(`spark-protobuf-assembly-**-SNAPSHOT.jar`) for the above tests and refactor the test cases to make them pass both SBT and Maven after this pr.

### Why are the changes needed?
Make connect server module can test pass using both SBT and maven.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
- Pass Github Actions
- Manual check 

```
build/mvn clean install -DskipTests
build/mvn test -pl connector/connect/server
```

all test passed after this pr.

### CODE REVIEW

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala [hunk 1]
```diff
@@ -3233,14 +3233,15 @@ class PlanGenerationTestSuite
     "connect/common/src/test/resources/protobuf-tests/common.desc"
 
   test("from_protobuf messageClassName") {
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala:3235]: If we deem this plan too intricate, we could delete or ignore test cases `from_protobuf messageClassName` and `from_protobuf messageClassName options` to temporarily abandon this testing scenario

File: connector/connect/server/src/test/java/org/apache/spark/sql/protobuf/protos/TestProto.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+/**
+ * SPARK-43646: This is generated by the `connector/connect/server/src/test/resources/test.proto`
+ * to test the spark protobuf uber jar. If you need to regenerate this file:
+ * 1. Modify `connector/connect/server/src/test/resources/test.proto` to generate Java files
+ * and replace the current file.
+ * 2. Replace all `com.google.protobuf` in the file with `org.sparkproject.spark_protobuf.protobuf.`
+ *  * <p>
+ *  * TODO(SPARK-44606): Generate this file and replace the package names in the file when testing.
+ */
+public final class TestProto {
```
Comment (LuciferYang) [inline connector/connect/server/src/test/java/org/apache/spark/sql/protobuf/protos/TestProto.java:15]: The task of generating Java files during testing and replacing all instances of `com.google.protobuf.` with `org.sparkproject.spark_protobuf.protobuf.` in the file will be carried forward as a follow-up.

**Comment by LuciferYang** [comment]
cc @hvanhovell @HyukjinKwon @zhengruifeng @rangadi and @pan3793

**Comment by HyukjinKwon** [comment]
I like this but cc @grundprinzip and @zhenlineo too

File: connector/connect/server/src/test/resources/test.proto [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ */
+syntax = "proto3";
+
+package org.apache.spark.sql.protobuf.protos;
+
+option java_multiple_files = true;
+option java_outer_classname = "TestProto";
+
+message TestProtoObj {
```
Comment (zhenlineo) [inline connector/connect/server/src/test/resources/test.proto:24]: I do not fully understand the purpose of this file? Can you add some docs here to explain why we need this file and why the 3 generated java files should be checked in?

Comment (LuciferYang) [reply connector/connect/server/src/test/resources/test.proto:24]: https://github.com/apache/spark/pull/42236#discussion_r1280118522

Thanks for your review, I have put the relevant explanation and TODO into the generated Java files.

```
/**
 * SPARK-43646: This is generated by the `connector/connect/server/src/test/resources/test.proto`
 * to test the spark protobuf uber jar. If you need to regenerate this file:
 * 1. Modify `connector/connect/server/src/test/resources/test.proto` to generate Java files
 * and replace the current file.
 * 2. Replace all `com.google.protobuf` in the file with `org.sparkproject.spark_protobuf.protobuf.`
 *  * <p>
 *  * TODO(SPARK-44606): Generate this file and replace the package names in the file when testing.
 */
```

In addition to generating Java files from proto, the content of the generated files also needs path replacement. I haven't figured out how to automate this task yet, so I have left it as a TODO. For this TODO, I have added both the proto files and the generated Java files in the current PR. The generated Java files are the ones to be used currently, and the proto files are to be used for the TODO.

**Comment by LuciferYang** [comment]
> Would it be easier if we change maven to use the unshaded jar?

https://github.com/apache/spark/blob/5c36c58047724885864cb781f17038a6b9c94513/connector/protobuf/pom.xml#L95-L100

Since `shadedArtifactAttached` is configured to be true, Maven will no longer publish the unshaded jar. Following Spark's previous practice, if both unshaded jars and shaded jars are needed, then it needs to be split into two modules, such as `kafka-0-10` and `kafka-0-10-assembly`, `kinesis-asl` and `kinesis-asl-assembly`. So, if we need Maven to publish the unshaded jar, we need to split the `protobuf` module into two modules: `protobuf` and `protobuf-assembly`. I had a previous PR that did this work: https://github.com/apache/spark/pull/41466.`

**Comment by LuciferYang** [comment]
While I'm not certain if it's reasonable, I still want to point out that relocating the content of the `spark-protobuf` module may result to a poorer user experience: In order to use this sql function, users have no choice but to relocate the content of the Java PB description files used in their business according to Spark's project rules. Is this really user-friendly for existing data and businesses? Meanwhile, the `spark-protobuf` module is a module that won't be packaged into spark-client tar ball, is it very risky to only publish unshaded jars? @rangadi @HyukjinKwon

File: connector/connect/server/src/test/resources/test.proto (hunk 3 above)
Comment (LuciferYang) [reply connector/connect/server/src/test/resources/test.proto]: @zhenlineo " have deleted this file in this pr, and I will add it back once the TODO is completed.

**Comment by pan3793** [comment]
Agree with @LuciferYang's opinion, better to use vanilla protobuf here.

Generally, the relocation technically should be used carefully, and should not affect user API. Some users may look for examples from test cases, it should be consistent with the end users' code. (if someone copies code snippets from test code but can not run with their own project, really a bad experience)

Historically, Hadoop and ORC have transitive protobuf 2.5 deps, which may cause deps conflict without shading and relocation in Spark, but it's not an issue today.

**Comment by LuciferYang** [comment]
Yes, if we accept that the `spark-protobuf` module does not perform `relocation` for `protobuf`, then this issue will become simpler and the cost for users will be relatively lower.

**Comment by LuciferYang** [comment]
cc @srowen This pr should be able to solve the test failure you mentioned in the dev email, I would greatly appreciate if you could participate in reviewing it, thanks ~

On the other hand, I am wondering about your opinion on https://github.com/apache/spark/pull/42236#issuecomment-1664828251

**Comment by LuciferYang** [comment]
@srowen No, this is not the impact brought by this pull request, it's just the current state of the `spark-protobuf` module.

The purpose of this pull request is to make both sbt and maven pass the test under the current state of the `spark-protobuf` module.

The current state of the `spark-protobuf` module mentioned here is similar to what you understood, with specific details similar to the comments in the following function：

https://github.com/apache/spark/blob/433fb2af8a3ca239958cb7b006e2924ecfac0d56/connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/functions.scala#L118-L136

This is unrelated to Spark Connect itself, Spark Connect just proxies this functions

**Comment by srowen** [comment]
OK if this doesn't change the current state, just resolves a test failure, then seems fine.

**Comment by LuciferYang** [comment]
> > While I'm not certain if it's reasonable, I still want to point out that relocating the content of the `spark-protobuf` module may result to a poorer user experience: In order to use this sql function, users have no choice but to relocate the content of the Java PB description files used in their business according to Spark's project rules(replcace all `com.google.protobuf.` to`org.sparkproject.spark_protobuf.protobuf.` in their generated Java files, Perhaps I misunderstood something?). Is this really user-friendly for existing data and businesses? Meanwhile, the `spark-protobuf` module is a module that won't be packaged into spark-client tar ball, is it very risky to only publish unshaded jars? @rangadi @HyukjinKwon
> 
> +1 for @LuciferYang 's opinion. It's not reasonable that user has to package its protobuf jar relocated. Since the spark-protobuf jar is not included in the spark-client binary, I think a better solution would be publishing both unshaded and shaded jars. Users could simply packing their protobuf message classes into a jar and use spark-protobuf jar as needed. The shaded spark-protobuf jar is kept for compatibility and also for last resort.
> 
> According to https://protobuf.dev/support/cross-version-runtime-guarantee/, I believe as long as Spark uses higher 3.x versions, user compiled java classes should be fine to run.

There was already a PR attempt to split the `spark-proto` module into two modules, `spark-proto` and `spark-assembly: https://github.com/apache/spark/pull/41466

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala (hunk 1 above)
Comment (rangadi) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala:3235]: This is great! Thank you.

File: connector/connect/server/src/test/java/org/apache/spark/sql/protobuf/protos/TestProto.java (hunk 2 above)
Comment (rangadi) [reply connector/connect/server/src/test/java/org/apache/spark/sql/protobuf/protos/TestProto.java:10]: Could you update the doc comment to say why this is included (the current comment talks about a fix, but not the problem). Looks like this uses shaded version of protobuf library.

**Review by rangadi — APPROVED** [review]
LGTM. Thanks for doing this. Hopefully we can shade the java classes at build time. Thanks for handling this tricky issue with using Java classes.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala [hunk 4]
```diff
@@ -3233,14 +3233,15 @@ class PlanGenerationTestSuite
...
-    binary.select(pbFn.from_protobuf(fn.col("bytes"), classOf[StorageLevel].getName))
+    binary.select(
+      pbFn.from_protobuf(fn.col("bytes"), "org.apache.spark.sql.protobuf.protos.TestProtoObj"))
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala:3237]: After https://github.com/apache/spark/pull/41153, protobuf 2.5 is no longer in classpath :)

**Comment by LuciferYang** [comment]
> Could you update the doc comment to say why this is included (the current comment talks about a fix, but not the problem). Looks like this uses shaded version of protobuf library.

Thanks for your review @rangadi , I will add more information later.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala:3237]: @rangadi I'm sorry that I neglected the historical context of adding these functions. Spark 3.5 no longer depends on Protobuf 2.5, and these functions are currently marked as `@Experimental`. In this situation, is it possible for us to make some changes?

[the remark this answers is not in the document]
Comment (rangadi) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala:3237]: That requires removing shading of protobuf classes. It needs more discussion. In general, including non-shaded jars is problematic (that is how original protobuf 2.5 version jar remained unchanged for many years).

**Comment by LuciferYang** [comment]
> Could you update the doc comment to say why this is included (the current comment talks about a fix, but not the problem). Looks like this uses shaded version of protobuf library.

done

**Comment by LuciferYang** [comment]
What suggestions does everyone have on how to proceed with fixing this issue? Should we have it fixed before the next 3.5.0 RC release?

**Comment by HyukjinKwon** [comment]
What are the standing comments? Otherwise, I would guess its good to go?

**Comment by LuciferYang** [comment]
@HyukjinKwon Hmm... I think the fix should be fine. However, as Spark 3.5 no longer needs to support protobuf 2.5, I hope to make some changes to the `Experimental` Apis involved in this pr to make it no longer intrude into user code.

The related viewpoints include：

https://github.com/apache/spark/pull/42236#issuecomment-1664828251, https://github.com/apache/spark/pull/42236#issuecomment-1665167038, 
https://github.com/apache/spark/pull/42236#issuecomment-1668006357

Of course, we could also choose to fix this issue for Spark 3.5 using this pr first and have more discussions or changes for this in Spark 4.0.

**Comment by LuciferYang** [comment]
[4ba45e0](https://github.com/apache/spark/pull/42236/commits/4ba45e0de2ee0dcfb7bad3e8811b10e19a5885c3)  has automated the process of generating `.java` from `.proto` and replacing all instances of `com.google.protobuf.` with 'org.sparkproject.spark_protobuf.protobuf.' in the `.java` files. Thus, the TODO task SPARK-44606 is no longer needed. ""

**Comment by LuciferYang** [comment]
@HyukjinKwon @rangadi Since there are new code changes in  [4ba45e0](https://github.com/apache/spark/pull/42236/commits/4ba45e0de2ee0dcfb7bad3e8811b10e19a5885c3), could you please take another look? If there are no new issues, I will merge this pr if GA passes.

**Comment by LuciferYang** [comment]
GA passed and I have manually verified with Maven, the tests can pass.

**Comment by HyukjinKwon** [comment]
@LuciferYang BTW, is this a real blocker? or test-only issue?

**Comment by LuciferYang** [comment]
> @LuciferYang BTW, is this a real blocker? or test-only issue?

hmm... If we accept that Maven tests inevitably fail, then this is not a blocker

**Comment by LuciferYang** [comment]
@HyukjinKwon So I don't think this is a real blocker, but we should at least make the maven test pass in master/branch-3.5, even if it means removing these two cases.

**Comment by LuciferYang** [comment]
I will merge this pr today If there are no more comments  @rangadi @HyukjinKwon  @hvanhovell @zhengruifeng @rangadi @pan3793 @grundprinzip @zhenlineo

**Comment by LuciferYang** [comment]
Merged into master and branch-3.5 to fix the maven test for the connect-server module

**Comment by LuciferYang** [comment]
Thanks all

**Comment by MaxGekk** [comment]
After this changes, I cannot compile Spark as a maven project in IntelliJ IDEA anymore. It fails with the error (see the screenshot). Any idea how to fix the issue?
```
java: package org.sparkproject.spark_protobuf.protobuf does not exist
```
Screenshot 2023-08-31 at 10 02 24

**Comment by LuciferYang** [comment]
@MaxGekk I thought for a while, I have no better way to solve this issue, indeed, it's impossible to be perfect. I have decided to revert this pr and ignore the corresponding two test cases.

**Pull request closed without merging**